### PR TITLE
RRF-87: Search Autocomplete

### PR DESCRIPTION
**Pull request by arianaantonio**
- Added temp API config
- Hooked up autocomplete and search to live data
- Added "brand" to RCHSearchProduct in SDK

### CODE REVIEW

**Comment by KingOfBrian** [comment]
I'm getting `Error getting current user` when going to that screen. There's a chance that there's some UserDefaults configuration that you have set up that is having it work for you. Can you try a fresh install and see if it works? The Client configuration workaround isn't getting triggered because it's missing.

File: Demo/RCHSearchViewController.swift [hunk 1]
```diff
@@ -19,8 +19,8 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
     @IBOutlet weak var searchProductsImageView: UIImageView!
     @IBOutlet weak var autocompleteTableView: UITableView!
     
-    var productArray: [RCHRecommendedProduct] = []
-    var autocompleteArray: [String] = ["Boots", "Bomber", "Boucle", "Boyfriend"]
+    var productArray: [RCHSearchProduct] = []
+    var autocompleteArray: [String] = []
```
Comment (KingOfBrian) [inline Demo/RCHSearchViewController.swift]: Don't include the type in the variable name. I'd use something like `products` and `autocompleteSuggestions` #pp

File: Demo/RCHSearchViewController.swift [hunk 2]
```diff
@@ -170,18 +205,18 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
...
+                let searchString = searchTerm.lowercased()
                 let highlightColor = UIColor(red: 0, green: 121/255, blue: 253/255, alpha: 1)
                 let blueAttribute = [NSBackgroundColorAttributeName : highlightColor]
-                
-                let attributedString = NSMutableAttributedString(string: autocompleteString, attributes: nil)
-                attributedString.addAttributes(blueAttribute, range: NSRange.init(location: 0, length: searchTermLength))
-                
+                let attributedString = NSMutableAttributedString(string: autocompleteString)
+            
+                let range: Range<String.Index> = autocompleteString.range(of: searchString)!
```
Comment (KingOfBrian) [inline Demo/RCHSearchViewController.swift]: I got a crash on this a few times, not entirely sure what I was doing differently #req

Comment (KingOfBrian) [reply Demo/RCHSearchViewController.swift]: Typing fast seems to crash it. I think `searchString` is updated faster than the response.

**Comment by KingOfBrian** [comment]
You have a merge conflict @arianaantonio ! Also, The status bar should be updated to white inside the search app.

**Comment by KingOfBrian** [comment]
Looks good @arianaantonio !
One thing we should probably do is put a 0.5 second delay on triggering both requests so we don't bog the system down. These can be triggered via timer in didSet on `searchTerm`.

 A few polish suggestions to take the pattern we talked about further: (#pp)
- Order the methods / vars -- State Management, API Calls, IBActions, UI Delegates.
  - I like to have my IBActions and UI Delegates just modify VC state or trigger API calls, and not have any conditionals or direct view modification. This isn't an absolute, a few small things is fine, but if things get bigger, it's best to model.
- Have the `searchTerm` property manage the text in the `UISearchBar`
- Have the attributed string returned in a method like `attributedText(for row: Int) -> NSAttributedString`

File: Demo/RCHSearchViewController.swift [hunk 3]
```diff
@@ -19,30 +19,56 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
...
+                searchProductsView.isHidden = false
+            } else {
+                autocompleteTableView.isHidden = false
+            }
+        }
+    }
+    var searchTerm = "" {
+        didSet {
+            searchBar.text = searchTerm
+            var _ = Timer.scheduledTimer(timeInterval: 0.5, target: self, selector: #selector(self.searchForProducts), userInfo: nil, repeats: false)
```
Comment (KingOfBrian) [inline Demo/RCHSearchViewController.swift]: We also want to save the timer and cancel the last one that was submitted so we don't just delay the search, but only do one search 0.5 seconds after the last keypress.

File: Demo/RCHSearchViewController.swift [hunk 4]
```diff
@@ -52,94 +78,109 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
         autocompleteTableView.tableFooterView = footerView
         
         searchResultsCollectionView!.register(UINib(nibName: "RCHProductCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: reuseIdentifier)
+        
+        configureAPI()
     }
     
-    @IBAction func handleTapOnFooter(sender: UITapGestureRecognizer) {
-        resetSearch()
+    func resetSearch() {
```
Comment (KingOfBrian) [inline Demo/RCHSearchViewController.swift]: `dismissSearch` may be a better name.

File: Demo/RCHSearchViewController.swift [hunk 5]
```diff
@@ -52,94 +78,109 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
...
+        RCHSDK.defaultClient().sendRequest(autocompleteBuilder.build(), success: { (responseObject) in
+            
+            guard let responseAutocompleteSuggestions = responseObject as? [RCHAutocompleteSuggestion] else {
+                print("Result Error")
+                return
+            }
+            
+            self.autocompleteSuggestions = responseAutocompleteSuggestions.map({$0.text!})
+        }) { (responseObject, error) in
+            print(error)
```
Comment (KingOfBrian) [inline Demo/RCHSearchViewController.swift:149]: We should present an alert if there's an error with the `localizedDescription` of the error as the message.

File: Demo/RCHSearchViewController.swift [hunk 6]
```diff
@@ -52,94 +78,109 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
...
-        searchForProducts(withTerm: searchText)
-        autocompleteTableView.reloadData()
-        print(searchText)
+        if searchBar.text == "" {
+            products.removeAll()
+            autocompleteSuggestions.removeAll()
+        } else {
+            searchTerm = searchText
+            autocomplete(withQuery: searchText)
+        }
```
Comment (KingOfBrian) [inline Demo/RCHSearchViewController.swift]: This entire behavior can get pushed into `searchTerm.didSet` I think.

File: Demo/RCHSearchViewController.swift [hunk 7]
```diff
@@ -52,94 +78,109 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
...
     }
     
     func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
-        guard let searchTerm = searchBar.text else {
+        guard let searchText = searchBar.text else {
             print("Error: no search term entered")
             return
         }
-        searchForProducts(withTerm: searchTerm)
+        searchTerm = searchText
```
Comment (KingOfBrian) [inline Demo/RCHSearchViewController.swift]: Since we're searching as we go, I don't think we have to update the searchTerm here.

**Review by KingOfBrian — CHANGES_REQUESTED** [review]
Few comments, I think the timer is the only #req item. Looks great!

File: Demo/RCHSearchViewController.swift (hunk 6 above)
Comment (arianaantonio) [reply Demo/RCHSearchViewController.swift]: I tried this before but if I move autocomplete into searchTerm.didSet then it gets called when the user selects an autocomplete suggestion as the searchTerm. Which means the autocomplete table pops up over the search results, and we don't want that here.

File: Demo/RCHSearchViewController.swift [hunk 8]
```diff
@@ -150,22 +163,12 @@ extension RCHSearchViewController: UISearchBarDelegate, UICollectionViewDelegate
...
     func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
-        if searchBar.text == "" {
-            products.removeAll()
-            autocompleteSuggestions.removeAll()
-        } else {
-            searchTerm = searchText
-            autocomplete(withQuery: searchText)
-        }
+        searchTerm = searchText
+        autocomplete(withQuery: searchText)
```
Comment (KingOfBrian) [inline Demo/RCHSearchViewController.swift:171]: I think we might want to do autocomplete on a timer too.

**Pull request merged**